src/write.c

```diff
@@ -2093,7 +2135,36 @@ avifResult avifEncoderAddImageGrid(avifEncoder * encoder,
     if (encoder->extraLayerCount == 0) {
         addImageFlags |= AVIF_ADD_IMAGE_FLAG_SINGLE; // image grids cannot be image sequences
     }
-    return avifEncoderAddImageInternal(encoder, gridCols, gridRows, cellImages, 1, addImageFlags);
+    const avifResult res = avifEncoderAddImageInternal(encoder,
+                                                       gridCols,
+                                                       gridRows,
+                                                       cellImages,
+                                                       /*gainMapGridCols=*/gridCols,
+                                                       /*gainMapGridRows=*/gridRows,
+                                                       /*gainMapCellImages=*/NULL,
+                                                       1,
+                                                       addImageFlags);
+    return res;
+}
+
+avifResult avifEncoderAddImageGridInternal(avifEncoder * encoder,
+                                           uint32_t gridCols,
+                                           uint32_t gridRows,
+                                           const avifImage * const * cellImages,
+                                           uint32_t gainMapGridCols,
+                                           uint32_t gainMapGridRows,
+                                           const avifImage * const * gainMapCellImages,
+                                           avifAddImageFlags addImageFlags)
+{
+    avifDiagnosticsClearError(&encoder->diag);
+    if ((gridCols == 0) || (gridCols > 256) || (gridRows == 0) || (gridRows > 256) || (gainMapGridCols == 0) ||
+        (gainMapGridCols > 256) || (gainMapGridRows == 0) || (gainMapGridRows > 256)) {
+        return AVIF_RESULT_INVALID_IMAGE_GRID;
+    }
```

Comment:
Nit: It is better to ignore `gainMapGridCols` and `gainMapGridRows` when `gainMapCellImages` is null:

```
    if ((gridCols == 0) || (gridCols > 256) || (gridRows == 0) || (gridRows > 256)) {
        return AVIF_RESULT_INVALID_IMAGE_GRID;
    }
    if (gainMapCellImages &&
        ((gainMapGridCols == 0) || (gainMapGridCols > 256) || (gainMapGridRows == 0) || (gainMapGridRows > 256))) {
        return AVIF_RESULT_INVALID_IMAGE_GRID;
    }
```